### PR TITLE
feat(review): rebuild the sticky comment as a mission-control status board

### LINKED ISSUE #1909: feat(review): sticky comment redesign — mission-control layout with run history
## Summary

Rebuild the sticky comment renderer to the approved v5 "mission control" layout (**mock-4, section 1** in the #1905 gist — mock-1 is superseded where they differ). Part of #1905; depends on #1906 (state/matching), #1907 (rich outputs), #1908 (prompts).

## Problem

Current sticky leads with token accounting, uses double-nested `<details>`, duplicates inline-comment content, shows a fictional dollar figure under subscription billing, and has no delta/history awareness.

## Proposed Solution

Render order (see mock-4 for exact content):

1. Header: `## 🔎 Lintro Review · round N · commit <sha>`
2. Readiness pill (derived: open P1 -> ⛔ Blocked; else P2 -> ⚠️ Changes requested; else P3 -> 🟡 Nits only; else ✅ Ready) + delta line `✔ n resolved · n new · n unchanged`
3. **Summary**: one-liner + walkthrough bullets (from #1907); bullets tied to open P1/P2 findings get severity dot + bold
4. **Why it's blocked/flagged**: model reasoning paragraphs + rubric fine-print + `Files needing attention:`
5. Severity tiles as a compact markdown table: blockers / warnings / nits / **fixed**
6. **Open findings** table: Δ · sev · finding · where · since (titles only, one line each)
7. **⚡ Fix-all prompt** panel (from #1908) labeled "all N still-open findings (rounds 1–N)", scope restated in the prompt's first line
8. **✔ Resolved** table: sev · ~~finding~~ · fixed-in `<sha>` · round
9. *This run* badges, two lines: `model · est. cost · tokens in · tokens out` then `transport · depth · files · checks · duration`
10. `---` then ONE collapsible `🕘 Run history — N runs · ~$x · ~xk tokens · N models`: cumulative badges (models ×counts first, est. cost, tokens totals, findings raised·resolved) + per-run table (run · commit · verdict · model · open · fixed · tokens in/out · est. cost · duration) + prior-round mini-summary lines
11. One-line footer

**Cut vs earlier draft:** files-reviewed table; cumulative section outside the collapsible; previous-runs as a second collapsible; 1–5 confidence score.

**Degraded path:** when inline posting fails for any finding, render a warning row above the Open table (count + cause) and fold those findings' full details into the sticky until inline posting succeeds.

Rules: no cost figure presented as billed — `transport: cli · subscription` badge + `~` prefix carry the honesty; no nested `<details>` deeper than one level; comment (incl. state blob) < 65,536 chars with oldest-history pruning.

## Implementation Notes

- `lintro/ai/review/github_sticky.py` — full renderer rewrite; keep the marker comment (`<!-- lintro-ai-review -->`) for upsert.

## Benefits

- Verdict-first living status board; zero duplication with inline comments; honest billing; visible history.

Part of #1905.

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Commit Summary (Conventional Commits)

- Title (required, present tense):

  ```text
  feat(review): rebuild the sticky comment as a mission-control status board
  ```

- Type:
  - [x] feat (minor)
  - [ ] fix / perf (patch)
  - [ ] docs
  - [ ] refactor
  - [ ] test
  - [ ] chore / ci / style

- Breaking change:
  - [ ] `!` in title or `BREAKING CHANGE:` footer included

### Release Trigger Rules (exact)

- A merged PR will bump the version based on its title (squash merge required):
  - `feat(...)` or `feat:` → MINOR bump
  - `fix(...)` / `fix:` or `perf(...)` / `perf:` → PATCH bump
  - Any title with `!` after the type (e.g. `feat!:` or `feat(scope)!:`) or a body
    containing `BREAKING CHANGE:` → MAJOR bump
- Use squash merge so the PR title becomes the merge commit title.

## What’s Changing

Rebuilds `lintro/ai/review/github_sticky.py` to the approved v5 "mission
control" layout (epic #1905, mock-4 section 1). The sticky comment becomes an
**index + status board** instead of a telemetry dump that duplicated every
inline comment.

Render order:

1. `## 🔎 Lintro Review · round N · commit <sha>`
2. deterministic readiness pill (`⛔ Blocked` / `⚠️ Changes requested` /
   `🟡 Nits only` / `✅ Ready`) plus the delta line
   `✔ n resolved · n new · n unchanged since round N-1`
3. **Summary** — headline plus walkthrough bullets; a bullet tied to an open
   P1/P2 finding carries that severity dot and is bolded, so a blocker cannot
   read as neutral prose
4. **Why it's blocked/flagged** — the model's reasoning paragraphs, the verdict
   rubric as fine-print, and `Files needing attention:`
5. severity tiles (`blockers / warnings / nits / fixed`)
6. **Open findings** table — `Δ · Sev · Finding · Where · Since`, titles only
7. **⚡ Fix-all prompt** panel scoped to all still-open findings (rounds 1–N)
8. **✔ Resolved** table with the fixing commit and round
9. *This run* badges, two lines, model-first ordering
10. `---` then exactly **one** `🕘 Run history` collapsible: cumulative badges,
    the per-run table, and one mini-summary line per prior round
11. one-line footer

### Honesty and structure rules encoded

- No figure is presented as billed: the `transport: cli · subscription` badge
  and the `~` prefix carry that, and the fictional dollar headline is gone.
- **No nested `<details>`** anywhere — asserted by a test that walks the
  rendered body and checks the collapsible depth never exceeds 1.
- The comment always fits GitHub's 65,536-char cap, pruning **oldest run
  history first**, then resolved findings, then open findings — each stage
  leaving a visible marker rather than dropping content silently.

### Degraded path

New `InlinePostFailure` model. When inline posting fails, those findings have
no surface at all, so the sticky renders a warning row above the open table
(count + cause) and folds their full detail back in — flattened, to keep the
no-nesting rule. `post_review_to_github` re-renders and updates the sticky in
place against the unchanged prior state, so the round is never double-counted
and no second sticky is ever posted.

## Checklist

- [x] Title follows Conventional Commits
- [x] Tests added/updated
- [x] Docs updated if user-facing
- [x] Local CI passed (`./scripts/local/run-tests.sh`)

## Closes

- Closes #1909
- Relates to #1906

## Details

- New `tests/unit/ai/review/test_github_sticky_mission_control.py` (22 tests)
  covers the readiness-pill derivation (including that questions never move
  it), the delta line, the Δ/since columns, resolved provenance,
  severity-marked summary bullets, the single-level-details rule, the degraded
  path, table-cell pipe escaping, and the size-cap pruning order.
- The history-pruning test sweeps a range of finding counts rather than
  hard-coding the count at which the cap bites, and asserts the invariants at
  each: the comment always fits, the rounds still shown are always the newest
  contiguous block, any drop is announced, and no finding is sacrificed while
  history remains to shed.
- Six tests in `test_github.py` asserted the old layout (`Cumulative (this
  PR):`, `Previous runs (N)`, the section-aware `_cap_body`) and were rewritten
  against the new contract. `_cap_body` is now only the final hard-truncation
  safety net; section-aware pruning moved into `_fit_body`.
- `github_constants.py`: added `STICKY_FOOTER` and `SHORT_SHA_LENGTH`, removed
  the five regexes that only served the old section-eliding capper.

### CODE REVIEW

File: lintro/ai/review/github_sticky.py [hunk 1]
```diff
@@ -92,67 +233,951 @@ def build_sticky_comment(
...
+    lower, upper = 0, _PRUNE_SEARCH_CEILING
+    while lower <= upper:
+        middle = (lower + upper) // 2
+        candidate = assemble(limits=replace(limits, **{field: middle}))
+        if len(candidate) <= MAX_COMMENT_CHARS:
+            best = candidate
+            lower = middle + 1
+        else:
+            upper = middle - 1
+    return best
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/review/github_sticky.py:405]: Fixed in 6ec2030d. `_largest_fitting` now takes a `ceiling` and starts the search at `min(ceiling, _PRUNE_SEARCH_CEILING)`; `_fit_body` receives `open_count`/`resolved_count` from the caller, which already computes them off `match.records`. Measurable: the review test module dropped from ~10.8s to ~5.2s.

File: lintro/ai/review/github_sticky.py [hunk 2]
```diff
@@ -92,67 +233,951 @@ def build_sticky_comment(
...
+            checklist_display=checklist_display,
+            question_map=question_map,
+        ),
+        render_agent_prompt_panel(
+            findings=open_findings,
+            scope=AgentPromptScope(
+                kind=AgentPromptScopeKind.ALL_OPEN,
+                round_number=round_number,
+            ),
+        ),
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/review/github_sticky.py:482]: Fixed in 6ec2030d. `_sorted_open_findings` now sorts by `normalize_file_path(finding.file)`, matching the key `_current_records` stores. Good catch — the divergence only showed once a pruning limit applied, which is exactly when the table and the fix-all prompt must not disagree about which findings they cover.

File: lintro/ai/review/github_sticky.py [hunk 3]
```diff
@@ -92,67 +233,951 @@ def build_sticky_comment(
...
+        The delta line, or an empty string on round 1 where there is nothing
+        to compare against.
+    """
+    if round_number <= 1:
+        return ""
+    unchanged = len(match.carried) + len(match.regressed)
+    return (
+        f"✔ {len(match.resolved)} resolved · **{len(match.new)} new** · "
+        f"{unchanged} unchanged since round {round_number - 1}"
+    )
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/review/github_sticky.py]: Fixed in 6ec2030d, along with two other reviewers who flagged the same contradiction. `_delta_line` now builds its parts list and appends `↩ N regressed` only when there are regressions, so the common case stays short and the line can no longer disagree with the `↩ regressed` cell in the table below it. Covered by `test_delta_line_reports_regressions_separately` and `test_delta_line_omits_the_regressed_clause_when_there_are_none`.

File: lintro/ai/review/github_sticky.py [hunk 4]
```diff
@@ -92,67 +233,951 @@ def build_sticky_comment(
...
+        model = run.model or "unknown"
+        counts[model] = counts.get(model, 0) + 1
+    return sorted(counts.items())
+
+
+def _fmt_compact(*, value: int) -> str:
+    """Format a large count compactly, for example ``24.9k``."""
+    if value < 1000:
+        return str(value)
+    return f"{value / 1000:.1f}k"
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/review/github_sticky.py:1286]: Fixed in 6ec2030d — added the millions branch. The value here is the cumulative token total across every retained run (up to 30), so seven figures is reachable on a long-lived PR. Covered by `test_history_summary_renders_millions_of_tokens_as_millions`.

File: lintro/ai/review/github.py [hunk 5]
```diff
@@ -126,10 +137,45 @@ def post_review_to_github(
...
+        # only ever *updated* — a failed lookup skips rather than posting a
+        # second sticky on the PR.
+        sticky_id = _sticky_comment_id(reporter=gh_reporter, known=comment_id)
+        if sticky_id is not None:
+            reporter_body = render(
+                inline_failure=InlinePostFailure(
+                    reason="the review API rejected the inline comments",
+                    findings=tuple(inline_findings),
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/review/github.py]: Fixed in 6ec2030d with the minimal wording change you suggested: the reason is now "the inline review comments could not be posted". `_post_inline_findings` only surfaces a boolean, so a 422, a 5xx and a timeout are indistinguishable here and the text must not name a cause the code never observed. Threading the real status out of `GitHubPRReporter.api_request` means changing shared reporter surface, which belongs to the inline-comment rebuild in #1911 — the renderer already accepts any reason string, so that lands without touching this code.

File: lintro/ai/review/github.py [hunk 6]
```diff
@@ -126,10 +137,45 @@ def post_review_to_github(
         question_map=prompt_questions,
     ):
         success = False
+        # Degraded path (#1909): those findings now have no surface at all, so
+        # re-render the sticky with their detail folded in. ``prior_state`` is
+        # unchanged, so this round is not double-counted, and the comment is
+        # only ever *updated* — a failed lookup skips rather than posting a
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/review/github.py:149]: Already covered — this reviewed an earlier commit. `post_review_to_github`'s degraded wiring has four tests in `test_github.py` as of 7a39f90c/c2b06fca:

- `test_failed_inline_post_folds_details_into_the_sticky` — drives the real function with `api_request` returning False and asserts two `update_issue_comment` calls, the folded detail, and that the round is not double-counted.
- `test_failed_inline_post_never_posts_a_second_sticky` — `find_issue_comment` returning None; asserts `update_issue_comment` is never called and only one sticky exists.
- `test_sticky_comment_id_short_circuits_on_a_known_id` / `..._relocates_a_just_created_comment` — both branches of `_sticky_comment_id` directly.
- `test_unmappable_findings_are_folded_in_without_any_failure` — the no-diff-line path added in c2b06fca.

File: lintro/ai/review/github_sticky.py [hunk 7]
```diff
@@ -92,67 +233,951 @@ def build_sticky_comment(
...
+        and not record.is_question
+        and record.severity is severity
+    )
+    noun = _plural(count=count, noun=_VERDICT_NOUNS[verdict])
+    return f"{label} — {count} open {noun}"
+
+
+def _delta_line(*, match: FindingMatchResult, round_number: int) -> str:
+    """Render the round-over-round finding delta.
+
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/review/github_sticky.py:554]: Fixed in 6ec2030d — same issue two other reviewers raised. Went with the explicit `↩ N regressed` clause rather than relabelling the bucket, so the wording matches the Δ column verbatim. The clause is omitted entirely when there are no regressions to keep the common line short.

File: tests/unit/ai/review/test_github_sticky_mission_control.py [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+) -> None:
+    """The healthy path never duplicates inline-comment detail."""
+    body = _body_only(body=build_sticky_comment(result=sample_review_result))
+
+    assert_that(body).does_not_contain("could not be posted as")
+    assert_that(body).does_not_contain("not posted inline")
+
+
+# --- size capping ------------------------------------------------------------
+
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply tests/unit/ai/review/test_github_sticky_mission_control.py:646]: Fixed in 6ec2030d, though not by widening the range — `range(50, 200)` is 150 full renders of a 60 KB body, and it would still be a guess. The test now **bisects** for the crossing point (history shrinks monotonically as findings grow, so the predicate is well-behaved), then asserts the invariants in a window around wherever it actually lands. That is ~10 renders instead of 20 and it cannot drift: a layout change moves the crossing and the bisection follows it.

File: lintro/ai/review/github_sticky.py [hunk 9]
```diff
@@ -92,67 +246,1003 @@ def build_sticky_comment(
...
+            auth_mode=auth_mode,
+        ),
+    ]
+    if checklist_display is ChecklistDisplay.ALL:
+        sections.append("\n".join(_format_checklist_appendix_markdown(result=result)))
+    history = _history_section(
+        runs=runs,
+        limit=limits.history,
+        resolved_total=total_resolved,
+    )
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/review/github_sticky.py:496]: Fixed in 6ec2030d. `unchanged` is now `len(match.carried)` alone and regressions get their own clause — your diagnosis of the cause was exact: the open table uses `match.outcome_for(record)` and distinguishes the two, while `_delta_line` was summing them.

File: tests/unit/ai/review/test_github_sticky_mission_control.py [hunk 10]
```diff
@@ -0,0 +1,720 @@
...
+    MAX_COMMENT_CHARS,
+    STATE_MARKER_PREFIX,
+    STICKY_MARKER,
+)
+from lintro.ai.review.github_sticky import (
+    _fit_body,
+    _RenderLimits,
+    build_sticky_comment,
+    parse_review_state_v2,
+)
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply tests/unit/ai/review/test_github_sticky_mission_control.py:32]: Taking the second option you offered — documenting the coupling — and declining the first, with reasons.

**`_fit_body`/`_RenderLimits`:** your own analysis is the argument for keeping this. As you note, the scenario "even a single open finding overflows" *cannot arise with real content* because every model string is length-capped inside the assembler. That is precisely why the public API can't reach it: a `build_sticky_comment` version of this test would pass no matter what the floor did. I tried exactly that first and CodeRabbit flagged it as vacuous, correctly. A test-only injection parameter on `build_sticky_comment` would put test scaffolding in production signature to avoid importing one private name — worse trade. An explanatory comment is now at the import site in 6ec2030d.

**`_sticky_comment_id`:** an earlier pass of this same reviewer explicitly asked for it ("Also add a unit test for `_sticky_comment_id` directly (known=None path calling find_issue_comment, known=42 short-circuit path)"). I'm keeping it, and its behaviour is *also* covered through `post_review_to_github` by two integration tests, so a rename that inlines it would be caught by more than an ImportError.

Worth noting this is the established pattern here rather than a new coupling: `test_github.py` on main already imports `_cap_body`, `_format_findings_section` and `_partition_findings` directly.
